src/sys/socket/mod.rs

```diff
@@ -392,53 +403,25 @@ impl<'a> Iterator for CmsgIterator<'a> {
     // although we handle the invariants in slightly different places to
     // get a better iterator interface.
     fn next(&mut self) -> Option<ControlMessage<'a>> {
-        let sizeof_cmsghdr = mem::size_of::<cmsghdr>();
-        if self.buf.len() < sizeof_cmsghdr {
+        if self.buf.len() == 0 {
+            // The iterator assumes that `self.buf` always contains exactly the
+            // bytes we need, so we're at the end when the buffer is empty.
             return None;
         }
-        let cmsg: &'a cmsghdr = unsafe { &*(self.buf.as_ptr() as *const cmsghdr) };
 
-        // This check is only in the glibc implementation of CMSG_NXTHDR
-        // (although it claims the kernel header checks this), but such
-        // a structure is clearly invalid, either way.
+        // Safe if: `self.buf` is `cmsghdr`-aligned.
+        let cmsg: &'a cmsghdr = unsafe { &*(self.buf[..mem::size_of::<cmsghdr>()].as_ptr() as *const cmsghdr) };
```

Comment:
How do you guarantee that self.buf is correctly aligned?

Reply:
The caller does. In this case its done by the `CmsgSpace`, which ensures alignment using a properly aligned member.

Reply:
Well, it's not the caller but whoever creates the iterator. Here, the buffer comes from `recvmsg` and is a `CmsgSpace` which guarantees alignment.